advanced_source/torch_script_parallelism.rst

```diff
@@ -0,0 +1,276 @@
+    # stacked and summed together, returning the combined result.
+    class LSTMEnsemble(torch.nn.Module):
+        def __init__(self, n_models):
+            super().__init__()
+            self.n_models = n_models
+            self.models = torch.nn.ModuleList([
+                BidirectionalRecurrentLSTM() for _ in range(self.n_models)])
+
+        def forward(self, x : torch.Tensor) -> torch.Tensor:
+            results = []
+            for model in self.models:
+                results.append(model(x))
+            return torch.stack(results).sum(dim=0)
+
+    # For a head-to-head comparison to what we're going to do with fork/wait, let's
+    # instantiate the model and compile it with TorchScript
+    ens = torch.jit.script(LSTMEnsemble(n_models=4))
+
+    # Normally you would pull this input out of an embedding table, but for the
+    # purpose of this demo let's just use random data.
+    x = torch.rand(T, B, C)
+
+    # Let's run the model once to warm up things like the memory allocator
+    ens(x)
+
+    x = torch.rand(T, B, C)
+
+    # Let's see how fast it runs!
+    s = time.time()
+    ens(x)
+    print('Inference took', time.time() - s, ' seconds')
+
+On my machine, this network runs in ``2.05`` seconds. We can do a lot better!
+
+Parallelizing Forward and Backward Layers
+-----------------------------------------
+
+A very simple thing we can do is parallelize the forward and backward layers
+within ``BidirectionalRecurrentLSTM``. For this, the structure of the computation
+is static, so we don't actually even need any loops. Let's rewrite the ``forward``
+method of ``BidirectionalRecurrentLSTM`` like so:
+
+.. code-block:: python
+
+        def forward(self, x : torch.Tensor) -> torch.Tensor:
+            # Forward layer - fork() so this can run in parallel to the backward
+            # layer
+            future_f = torch.jit.fork(self.cell_f, x)
+
+            # Backward layer. Flip input in the time dimension (dim 0), apply the
+            # layer, then flip the outputs in the time dimension
+            x_rev = torch.flip(x, dims=[0])
+            output_b, _ = self.cell_b(torch.flip(x, dims=[0]))
+            output_b_rev = torch.flip(output_b, dims=[0])
+
+            # Retrieve the output from the forward layer. Note this needs to happen
+            # *after* the stuff we want to parallelize with
+            output_f, _ = torch.jit.wait(future_f)
+
+            return torch.cat((output_f, output_b_rev), dim=2)
+
+In this example, ``forward()`` delegates execution of ``cell_f`` to another thread,
+while it continues to execute ``cell_b``. This causes the execution of both the
+cells to be overlapped with each other.
+
+Running the script again with this simple modification yields a runtime of
+``1.71`` seconds for an improvement of ``17%``!
+
+Aside: Visualizing Parallelism
+------------------------------
+
+We're not done optimizing our model but it's worth introducing the tooling we
```

Comment:
I would move this to after the section wherer you improve parallelism further.

Reply:
I wanted to put this here because it allows us to see a delta of what happens when we add ``fork`` and ``wait``, rather than just seeing one example